ci_server/server.py

```diff
@@ -32,7 +39,77 @@ def do_GET(self):
         self.wfile.write(bytes(file, "utf8"))
         # TODO: implement logic for serving files here
 
-    def do_POST(self):
-        # TODO: implement CI logic here
-        self.send_response(404)
+    # verify the signature of the message
+    def verify_signature(self, post_data):
+        sha_name, signature = self.headers['X-Hub-Signature-256'].split('=')
+        if sha_name != 'sha256':
+            return ERROR
+        # TODO define a secret token when creating the webhooks
+        local_signature = hmac.new(
+            os.getenv("secretToken").encode(),
+            msg=post_data,
+            digestmod=hashlib.sha256
+        ).hexdigest()
+        return not hmac.compare_digest(local_signature, signature)
+
+    # clone the branch related to the push event
+    def clone_branch(self, data):
+        # retrieve clone url and branch name
+        clone_url = data["repository"]["clone_url"]
+        branch = data["ref"].split("/")[-1]
+        return os.system(
+            f"cd branches; git clone --single-branch -b {branch} {clone_url}"
```

Comment:
You should probably use a shallow clone here as we do not need the git history. It would be something like `git clone --single-branch --depth 1 -b {branch} {clone_url} {name_of_folder}`

Reply:
Thank you for the remark, I'll have a look at it